verifier/services/reorg_detector_service.go

```diff
@@ -0,0 +1,701 @@
+		"chainSelector", r.config.ChainSelector,
+		"minBlock", r.latestFinalizedBlock,
+		"maxBlock", r.latestBlock,
+		"tailSize", len(r.tailBlocks))
+
+	return nil
+}
+
+// sendReorgNotification sends a reorg notification with the common ancestor block.
+func (r *ReorgDetectorService) sendReorgNotification(lcaBlockNumber uint64) {
+	status := protocol.ChainStatus{
+		Type:         protocol.ReorgTypeNormal,
+		ResetToBlock: lcaBlockNumber,
+	}
+
+	select {
+	case r.statusCh <- status:
+		r.lggr.Infow("Sent reorg notification",
+			"chainSelector", r.config.ChainSelector,
+			"type", status.Type.String(),
+			"resetToBlock", lcaBlockNumber)
+	default:
+		r.lggr.Warnw("Status channel full, dropping reorg notification",
+			"chainSelector", r.config.ChainSelector)
+	}
+}
+
+// sendFinalityViolation sends a finality violation notification.
+// No reset block is provided - finality violations require immediate stop and manual intervention.
+func (r *ReorgDetectorService) sendFinalityViolation(violatedBlock protocol.BlockHeader, finalizedBlockNum uint64) {
+	r.lggr.Errorw("FINALITY VIOLATION - sending notification",
+		"chainSelector", r.config.ChainSelector,
+		"violatedBlock", violatedBlock.Number,
+		"violatedHash", violatedBlock.Hash,
+		"finalizedBlock", finalizedBlockNum)
+
+	status := protocol.ChainStatus{
+		Type:         protocol.ReorgTypeFinalityViolation,
+		ResetToBlock: 0, // No safe reset point exists
+	}
+
+	select {
+	case r.statusCh <- status:
+		r.lggr.Errorw("Sent finality violation notification - chain reader will stop",
+			"chainSelector", r.config.ChainSelector,
+			"type", status.Type.String())
+	default:
+		r.lggr.Warnw("Status channel full, dropping finality violation notification",
+			"chainSelector", r.config.ChainSelector)
+	}
+}
+
+// finalityViolationError is returned when a reorg violates finality.
+type finalityViolationError struct {
+	violatedBlock              protocol.BlockHeader
+	newChainBlock              protocol.BlockHeader
+	finalizedNum               uint64
+	latestFinalizedBlockExists bool
+}
+
+func (e *finalityViolationError) Error() string {
+	if !e.latestFinalizedBlockExists {
+		return fmt.Sprintf("finality violation: finalized block %d not found in tail", e.finalizedNum)
+	}
+	return fmt.Sprintf("finality violation: finalized block hash %s does not match new chain hash %s at height %d",
+		e.violatedBlock.Hash, e.newChainBlock.Hash, e.finalizedNum)
+}
+
+// Close stops the reorg detector and closes the status channel.
+//
+// Behavior:
+// 1. Signals monitoring goroutine to stop via context cancellation
+// 2. Waits for goroutine to finish (blocks until clean shutdown)
+// 3. Closes status channel (readers will receive channel close signal)
+//
+// Thread-safety:
+// - Safe to call multiple times (subsequent calls are no-ops)
+// - Blocks until monitoring goroutine exits.
+func (r *ReorgDetectorService) Close() error {
+	r.mu.Lock()
+	if !r.running {
+		r.mu.Unlock()
```

Comment:
Echoing my comment from above, the usage of this mutex is surprisingly confusing and error-prone for just wanting to serialize Start and Close